zebra-chain/src/block/tests.rs

```diff
@@ -0,0 +1,134 @@
+use std::io::{Cursor, Write};
+
+use chrono::NaiveDateTime;
+use proptest::{
+    arbitrary::{any, Arbitrary},
+    prelude::*,
+};
+
+use crate::sha256d_writer::Sha256dWriter;
+
+use super::*;
+
+#[cfg(test)]
+impl Arbitrary for BlockHeader {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: ()) -> Self::Strategy {
+        (
+            any::<BlockHeaderHash>(),
+            any::<MerkleTreeRootHash>(),
+            any::<SaplingNoteTreeRootHash>(),
+            (0i64..4_294_967_296i64),
+            any::<u32>(),
+            any::<[u8; 32]>(),
+            any::<EquihashSolution>(),
+        )
+            .prop_map(
+                |(
+                    block_hash,
+                    merkle_root_hash,
+                    sapling_root_hash,
+                    timestamp,
+                    bits,
+                    nonce,
+                    equihash_solution,
+                )| BlockHeader {
+                    previous_block_hash: block_hash,
+                    merkle_root_hash: merkle_root_hash,
+                    final_sapling_root_hash: sapling_root_hash,
+                    time: Utc.timestamp(timestamp, 0),
+                    bits: bits,
+                    nonce: nonce,
+                    solution: equihash_solution,
+                },
+            )
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+#[test]
+fn blockheaderhash_debug() {
+    let preimage = b"foo bar baz";
+    let mut sha_writer = Sha256dWriter::default();
+    let _ = sha_writer.write_all(preimage);
+
+    let hash = BlockHeaderHash(sha_writer.finish());
+
+    assert_eq!(
+        format!("{:?}", hash),
+        "BlockHeaderHash(\"bf46b4b5030752fedac6f884976162bbfb29a9398f104a280b3e34d51b416631\")"
+    );
+}
+
+#[test]
+fn blockheaderhash_from_blockheader() {
+    let some_bytes = [0; 32];
+
+    let blockheader = BlockHeader {
+        previous_block_hash: BlockHeaderHash(some_bytes),
+        merkle_root_hash: MerkleTreeRootHash(some_bytes),
+        final_sapling_root_hash: SaplingNoteTreeRootHash(some_bytes),
+        time: DateTime::<Utc>::from_utc(NaiveDateTime::from_timestamp(61, 0), Utc),
+        bits: 0,
+        nonce: some_bytes,
+        solution: EquihashSolution([0; 1344]),
+    };
+
+    let hash = BlockHeaderHash::from(blockheader);
+
+    assert_eq!(
+        format!("{:?}", hash),
+        "BlockHeaderHash(\"738948d714d44f3040c2b6809ba7dbc8f4ba673dad39fd755ea4aeaa514cc386\")"
```

Comment:
This hash is generated from the current implementation rather than being an external source of truth, correct? It looks like block 415000 https://explorer.zcha.in/blocks/415000 wasn't very big, having only one transaction, so maybe we could add the entire block to the test vectors and check that the hash matches the expected one?

Reply:
Not sure about the easiest/fastest way to get the bytes of a particular block.

Reply:
Update: after looking through the Zebra source, I think it will be difficult/annoying to use Zebra itself to pull out the raw bytes of the block.  The problem is that the whole architecture is set up to convert things into internal representations from bytes at the edge, which works great after serialization is done, but works less well when it's not.

Reply:
@gtank points out that this can be done by `zcash-cli getblock 415000 false`

Reply:
Tracked in #218